### PR TITLE
Refactoring of `primitive_root`

### PR DESCRIPTION
**Pull request by haru-44**
p into two patterns and made them independent as a function.

* `_primitive_root_prime_iter`
Generates the primitive roots of `p`. Once one primitive root `g` is found, the other primitive roots are represented by powers of `g`. In this case, the exponent and `p-1` are relatively prime. Therefore, there is no need to perform power calculations to determine the primitive roots every time.

* `_primitive_root_prime_power_iter`
Generates the primitive roots of `p**e`. Let `g` (`0 < g < p`) be the primitive root of `p`. If `pow(g+s*p,p-1,p**2)!=1`, then `g+s*p` is primitive root of `p**e`. There is always one integer `s` (`0 <= s < p`) per `g` that does not satisfy this condition, and the other `p-1` are primitive roots. Thus, we find `totient(p)*(p-1)` primitive roots of `p**2`, which is consistent with `totient(totient(p**2))`. The primitive roots of `p**3` are all numbers congruent to the primitive roots of `p**2` by `p`. The same is true for `p**e` (`e > 3`), which is consistent with the `totient(totient(p**e))`. From the above, we see that `g+(k+m*p)*p` is all of the primitive roots of `p**e`. Where `g` is primitive root of `p` (`0<g<p`), `0 <= k < p`, `0 <= m < p**(e-2)`, `g**(p-1)+(p-1)*g**(p-2)*k*p % p**2 != 1`.

* `_primitive_root_prime_power2_iter`
Generates the primitive roots of `2*p**e`. If `g` is the primitive root of `p**e`, then the odd one of `g` and `g+p**e` is the primitive root of `2*p**e`. Since `totient(totient(2*p**e))=totient(totient(p**e))`, this is all the primitive roots of `2*p**e`.

#### Release Notes

* ntheory
  * Added `smallest` option to `primitive_root`

### CODE REVIEW

**Comment by sylee957** [comment]
I'm not sure if `ordered` keyword is needed everywhere.
We should do that only for the first or last time, and if it is not relevant for user function, it's better not to have that feature

**Comment by haru-44** [comment]
thank you for comment.
deleted `ordered` keyword.

**Comment by smichr** [comment]
I wonder if it might be better still to leave the sorting for the public facing functions so the low-level can just return what's convenient and the higher level functions sort it if necessary. As long as it is canonical it need not be sorted for purposes of testing.

**Comment by smichr** [comment]
Please consider whether sorting is necessary at this level.

**Comment by haru-44** [comment]
To begin with, `primitive_root` is designed to return the smallest primitive root.
If a low-level function outputs without regard to order, `primitive_root` can only return an answer after getting all primitive roots, which is an exaggeration for getting one primitive root.

So how about the following implementation?

* Low-level functions output without regard to order.
* The `primitive_root` optionally allows the user to choose whether to return the smallest primitive root or one primitive root of whatever.
* If it returns one primitive root, it calls a low-level function to find the primitive root fast.
* If returning the smallest of primitive roots, check if it is a primitive root from the smaller one, as in the current case.

**Comment by smichr** [comment]
> If it returns one primitive root, it calls a low-level function to find ~the~ a primitive root fast.
If returning the smallest of primitive roots, ~check if it is a primitive root from the smaller one, as in the current case.~ use `min` instead of `next`.

I think that would be ok -- I don't work with ntheory much, however, and don't know what the preferences of someone would be.

File: sympy/ntheory/residue_ntheory.py [hunk 1]
```diff
@@ -86,33 +94,110 @@ def _primitive_root_prime_iter(p):
...
+        elif p == 7 or p % 7 == 3:
+            # 3 is the smallest primitive root of p = 7,17,31
+            g = 3
+        else:
+            # 2 is the smallest primitive root of p = 5,11,13,19,29,37
+            g = 2
+    else:
+        v = [(p - 1) // i for i in qs]
+        for g in range(2, p):
+            if all(pow(g, pw, p) != 1 for pw in v):
```
Comment (smichr) [inline sympy/ntheory/residue_ntheory.py:114]:
```suggestion
    if p < 41:
        # small case
        if p == 23:
            g = 5
        elif p == 7 or p % 7 == 3:
            # 3 is the smallest primitive root of p = 7,17,31
            g = 3
        else:
            # 2 is the smallest primitive root of p = 5,11,13,19,29,37
            g = 2
    else:
        v = [(p - 1) // i for i in (factorint(p - 1)]
        for g in range(2, p):
            if all(pow(g, pw, p) != 1 for pw in v):
```

File: sympy/ntheory/residue_ntheory.py [hunk 2]
```diff
@@ -86,33 +94,110 @@ def _primitive_root_prime_iter(p):
...
-            yield a
-        a += 1
+            yield g + p**e
 
 
-def primitive_root(p):
+def primitive_root(p, smallest=True):
     """
-    Returns the smallest primitive root or None.
+    Returns the primitive root or None.
```
Comment (smichr) [inline sympy/ntheory/residue_ntheory.py:185]:
```suggestion
    Returns a primitive root of p or None.
```

File: sympy/ntheory/residue_ntheory.py (hunk 1 above)
Comment (smichr) [reply sympy/ntheory/residue_ntheory.py:114]: What if the multiplicity is more than one? Do we have to do the following, e.g. for p = 73 whose p-1 factors are {2:3, 3:2}:

```python
[(p - 1) // i**m for i,m in factorint(p - 1).items()]
```

Comment (smichr) [reply sympy/ntheory/residue_ntheory.py:114]: localizing factoring to where it is needed

File: sympy/ntheory/residue_ntheory.py [hunk 3]
```diff
@@ -122,10 +206,27 @@ def primitive_root(p):
...
-        return 1
+    if p < 2:
+        raise ValueError("p should be an integer greater than 1")
+    if p <= 4:
+        return p - 1
+    if not smallest:
+        t = trailing(p)
+        if t > 1:
+            return None
+        q = p >> t
```
Comment (smichr) [inline sympy/ntheory/residue_ntheory.py:217]:
```suggestion
        if p%2:
            q = p  # p is odd
        elif p%4:
            q = p//2  # p had 1 factor of 2
        else:
            return None  # p had more than one factor of 2
```

File: sympy/ntheory/residue_ntheory.py [hunk 4]
```diff
@@ -122,10 +206,29 @@ def primitive_root(p):
...
+        if isprime(q):
+            e = 1
+        else:
+            m = perfect_power(q)
+            if not m:
+                return None
+            q, e = m
+            if not isprime(q):
+                return None
+        if t:
```
Comment (smichr) [inline sympy/ntheory/residue_ntheory.py:229]:
```suggestion
        p_even = p%2 == 0
        if not p_even:
            q = p  # p is odd
        elif p%4:
            q = p//2  # p had 1 factor of 2
        else:
            return None  # p had more than one factor of 2
        if isprime(q):
            e = 1
        else:
            m = perfect_power(q)
            if not m:
                return None
            q, e = m
            if not isprime(q):
                return None
        if p_even:
```

**Comment by smichr** [comment]
If the changes look ok to you, this is ready.

**Comment by haru-44** [comment]
thank you.
I corrected one point, but everything else is fine.

**Pull request merged**